src/main.ts

```diff
@@ -1,127 +1,192 @@
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                event.preventDefault();
+                if (event.key != 'g') {
+                    keyArray = [];
+                } else {
+                    keyArray.push(event.key);
+                    if (keyArray.length === 2) {
+                        leaf.previewMode.applyScroll(0);
+                        keyArray = [];
+                    }
+                }
+            }
+        }
+        addEventListener('keydown', jumpTopEvent, { capture: true });
+
+        // Jump to bottom
+        //  1st evt registers g when CapsLock is on
+        //  2nd evt registers Shift+g
+        this.navScope.register([], 'g', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none' &&
+                evt.key === 'G'
+            ) {
+                this.jumpBottom(leaf);
+                return false;
+            }
+            return true;
+        });
+        this.navScope.register(['Shift'], 'g', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                this.jumpBottom(leaf);
+                return false;
+            }
+            return true;
+        });
+
+        app.keymap.pushScope(this.navScope);
+        console.log('Vim Reading View Navigation loaded.');
+    }
+
+    displayValue(leaf: MarkdownView): string {
+        return leaf.contentEl
+            .getElementsByClassName('markdown-reading-view')[0]
+            .getElementsByClassName('document-search-container')[0]
+            .getCssPropertyValue('display');
+    }
+
+    async onunload() {
+        app.keymap.popScope(this.navScope);
+        removeEventListener('keydown', this.jumpTopEvent, { capture: true });
+        console.log('Vim Reading View Navigation unloaded.');
+    }
+
+    getScroll(leaf: MarkdownView): number {
+        return leaf.previewMode.getScroll();
+    }
+
+    scrollDown(leaf: MarkdownView) {
+        const scroll = this.getScroll(leaf);
+        leaf.previewMode.applyScroll(scroll + this.settings.scrollDifference);
+    }
+
+    scrollUp(leaf: MarkdownView) {
+        const scroll = this.getScroll(leaf);
+        leaf.previewMode.applyScroll(scroll - this.settings.scrollDifference);
+    }
+
+    jumpBottom(leaf: MarkdownView) {
+        let scroll = this.getScroll(leaf);
+        leaf.previewMode.applyScroll(scroll + this.settings.scrollDifference);
+        let newScroll = this.getScroll(leaf);
+
+        while (newScroll != scroll) {
+            scroll = this.getScroll(leaf);
```

Comment:
I think you could set `scroll = newScroll` here to avoid calling it twice.

The first time you just called it before, after that it gets called after you scrolled in the loop, which means also just before.

Reply:
Right, good catch!